### PR TITLE
Refactor of CSL parser

### PR DESCRIPTION
**Pull request by marcusbfs**
Hey @euonymos , @klntsky and @errfrom, can you please verify/comment on this draft PR? Thanks

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -194,6 +194,7 @@
                 nodePackages.prettier
                 fd
                 git
+                gmp
```
Comment (errfrom) [inline flake.nix:197]: Why do we need `gmp`?

File: code-gen/parse-csl/src/Csl/Gen/Pureness.hs [hunk 2]
```diff
@@ -0,0 +1,147 @@
+module Csl.Gen.Pureness (
```
Comment (errfrom) [inline code-gen/parse-csl/src/Csl/Gen/Pureness.hs:1]: I suggest we rename the type to `FunPurity` and move it to `Csl.Gen.Types.FunPurity`. 'Purity' should be a more conventional term in this context.

File: code-gen/parse-csl/src/Csl/Gen/Utils.hs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+      case pre of
+        [] -> wrapText T.lowerHead post
+        [a] -> toLower a : post
+        _ -> map toLower (init pre) <> [last pre] <> post
+  where
+    (pre, post) = span isUpper str
+
+filterMethods :: Class -> [Method]
+filterMethods (Class name ms)
+  -- CostModel is a special case: it looks like a mapping type, but isn't
```
Comment (errfrom) [inline code-gen/parse-csl/src/Csl/Gen/Utils.hs:124]: This comment should be fixed to also include `VotingProcedures`, which is not a normal map and should have its methods exposed.

**Review by errfrom — COMMENTED** [review]
Looks good so far! The next step would be to add an ADT for generator configuration, so that we can specify various parameters in one place (e.g. common / mutating / throwing methods, and various exceptions to the rules). It should effectively serve as an interface to the generator, which global instance will be supposed to change more frequently than the existing set of general rules.

File: code-gen/parse-csl/src/Csl/Gen/Utils.hs [hunk 4]
```diff
@@ -0,0 +1,230 @@
+module Csl.Gen.Utils (
```
Comment (errfrom) [inline code-gen/parse-csl/src/Csl/Gen/Utils.hs:1]: To improve modularity, consider extracting all domain-specific functionality to different modules, and only use `Csl.Gen.Utils` for very general functions like `withSemicolon` or `wrapText`. I would suggest putting those new modules into the `Csl.Gen.Lib` namespace.

File: flake.nix (hunk 1 above)
Comment (marcusbfs) [reply flake.nix:197]: Without `gmp` I was getting an error during the `stack build` stage: 

```bash
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld.gold: error: cannot find -lgmp
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)
```

Adding `gmp` to the dependencies solved this issue.

File: code-gen/parse-csl/src/Csl/Gen/Pureness.hs (hunk 2 above)
Comment (marcusbfs) [reply code-gen/parse-csl/src/Csl/Gen/Pureness.hs]: Addressed in 275a386.

File: code-gen/parse-csl/src/Csl/Gen/Utils.hs (hunk 3 above)
Comment (marcusbfs) [reply code-gen/parse-csl/src/Csl/Gen/Utils.hs]: Addressed in 09b97f4.

File: code-gen/parse-csl/src/Csl/Gen/Utils.hs (hunk 4 above)
Comment (marcusbfs) [reply code-gen/parse-csl/src/Csl/Gen/Utils.hs:1]: Addressed in f2f7472.

File: flake.nix (hunk 1 above)
Comment (euonymos) [reply flake.nix:197]: Why do we use stack BTW?

File: code-gen/parse-csl/src/Csl/Gen/PureScript.hs [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    jsMethodName Method {..} = methodName name (fun'name method'fun)
+
+    psSig nullType m@Method {..} = trim $ unwords [if L.null argTys then "" else L.intercalate " -> " argTys <> " ->", resTy]
+      where
+        addTypePrefix pref x
+          | length (words x) > 1 = unwords [pref, "(" <> x <> ")"]
+          | otherwise = unwords [pref, x]
+
+        argTys = handleNumArgs $ (if not (isObj m) then id else (toType name :)) $ handleVoid True . arg'type <$> fun'args method'fun
```
Comment (klntsky) [inline code-gen/parse-csl/src/Csl/Gen/PureScript.hs:122]: Please keep line width <= 80

File: code-gen/parse-csl/src/Csl/Types.hs [hunk 6]
```diff
@@ -1,4 +1,11 @@
-module Csl.Types where
+module Csl.Types (
+  Arg (..),
```
Comment (klntsky) [inline code-gen/parse-csl/src/Csl/Types.hs:2]: Please use explicit exports/imports, see https://github.com/mlabs-haskell/styleguide

**Review by klntsky — CHANGES_REQUESTED** [review]
This is really good I think! I have two requests though

**Comment by klntsky** [comment]
Please also remove the `demo/` folder from the repo. It's no more needed.

File: code-gen/parse-csl/src/Csl/Gen/PureScript.hs (hunk 5 above)
Comment (marcusbfs) [reply code-gen/parse-csl/src/Csl/Gen/PureScript.hs]: Addressed in 5322ace.

**Comment by klntsky** [comment]
@marcusbfs please resolve the conflicts

File: flake.nix (hunk 1 above)
Comment (marcusbfs) [reply flake.nix:197]: Decision after meeting: leave it as it is for now

**Pull request merged**